### PR TITLE
gh-91922: Fix sqlite connection on nonstardard locales and paths

### LINKED ISSUE #91922: test_sqlite3 fails on non-UTF-8 locale
```
$ LC_ALL=en_US.iso88591 ./python -m test -vuall test_sqlite3
...
======================================================================
ERROR: test_ctx_mgr_rollback_if_commit_failed (test.test_sqlite3.test_dbapi.MultiprocessTests.test_ctx_mgr_rollback_if_commit_failed)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/serhiy/py/cpython/Lib/test/test_sqlite3/test_dbapi.py", line 1767, in test_ctx_mgr_rollback_if_commit_failed
    cx = sqlite.connect(TESTFN, timeout=self.CONNECTION_TIMEOUT)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe6 in position 16: unexpected end of data

======================================================================
ERROR: test_open_uri (test.test_sqlite3.test_dbapi.OpenTests.test_open_uri)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/serhiy/py/cpython/Lib/test/test_sqlite3/test_dbapi.py", line 674, in test_open_uri
    with managed_connect(TESTFN) as cx:
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/serhiy/py/cpython/Lib/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/home/serhiy/py/cpython/Lib/test/test_sqlite3/test_dbapi.py", line 44, in managed_connect
    cx = sqlite.connect(*args, **kwargs)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe6 in position 16: unexpected end of data

======================================================================
ERROR: test_open_with_path_like_object (test.test_sqlite3.test_dbapi.OpenTests.test_open_with_path_like_object)
Checks that we can successfully connect to a database using an object that
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/serhiy/py/cpython/Lib/test/test_sqlite3/test_dbapi.py", line 670, in test_open_with_path_like_object
    with managed_connect(path) as cx:
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/serhiy/py/cpython/Lib/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/home/serhiy/py/cpython/Lib/test/test_sqlite3/test_dbapi.py", line 44, in managed_connect
    cx = sqlite.connect(*args, **kwargs)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe6 in position 16: unexpected end of data

======================================================================
ERROR: test_trace_callback_content (test.test_sqlite3.test_hooks.TraceCallbackTests.test_trace_callback_content)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/serhiy/py/cpython/Lib/test/test_sqlite3/test_hooks.py", line 279, in test_trace_callback_content
    con1 = sqlite.connect(TESTFN, isolation_level=None)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe6 in position 16: unexpected end of data

----------------------------------------------------------------------
```

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Fix function sqlite.connect() and the sqlite.Connection constructor
  on non-UTF-8 locales.
* Fix support of bytes paths non-decodable with the current FS encoding.

Closes #91922

### CODE REVIEW

File: Lib/test/test_sqlite3/test_dbapi.py [hunk 1]
```diff
@@ -654,22 +652,58 @@ class OpenTests(unittest.TestCase):
...
+        path = FakePath(TESTFN)
         with managed_connect(path) as cx:
+            self.assertTrue(os.path.exists(path))
+            cx.execute(self._sql)
+
+    @unittest.skipIf(sys.platform == "win32", "skipped on Windows")
+    @unittest.skipIf(sys.platform == "darwin", "skipped on macOS")
+    @unittest.skipUnless(TESTFN_UNDECODABLE, "only works if there are undecodable paths")
+    def test_open_with_undecodable_path(self):
+        self.addCleanup(unlink, TESTFN_UNDECODABLE)
```
Comment (erlend-aasland) [inline Lib/test/test_sqlite3/test_dbapi.py:664]: I believe `managed_connect` will unlink, so this should not be necessary.

```suggestion
```

File: Lib/test/test_sqlite3/test_dbapi.py [hunk 2]
```diff
@@ -654,22 +652,58 @@ class OpenTests(unittest.TestCase):
...
+        with sqlite.connect(uri, uri=True) as cx:
+            with self.assertRaises(sqlite.OperationalError):
                 cx.execute(self._sql)
 
+    @unittest.skipIf(sys.platform == "win32", "skipped on Windows")
+    @unittest.skipIf(sys.platform == "darwin", "skipped on macOS")
+    @unittest.skipUnless(TESTFN_UNDECODABLE, "only works if there are undecodable paths")
+    def test_open_undecodable_uri(self):
+        uri = "file:" + urllib.parse.quote(TESTFN_UNDECODABLE)
+        self.addCleanup(unlink, TESTFN_UNDECODABLE)
```
Comment (erlend-aasland) [inline Lib/test/test_sqlite3/test_dbapi.py]: I believe `managed_connect` will unlink, so this should not be necessary.

```suggestion
```

**Comment by erlend-aasland** [comment]
Thanks, I'll review it tomorrow! From a quick glance, it seems like the main change is the `PyOS_FSPath` conversion, right?

**Comment by serhiy-storchaka** [comment]
No, the main change is calling `PySys_Audit()` before encoding the database path to `char *`.

`PyOS_FSPath()` is used instead of `PyUnicode_FSConverter()` in the module-level function to avoid unneeded encoding at that stage. In is not necessary change now (I thought it was necessary  at the beginning of work on this issue).

File: Modules/_sqlite/module.c [hunk 3]
```diff
@@ -77,11 +77,11 @@ pysqlite_connect_impl(PyObject *module, PyObject *database, double timeout,
     else {
         Py_INCREF(isolation_level);
     }
-    PyObject *res = PyObject_CallFunction(factory, "OdiOiOii", database,
+    PyObject *res = PyObject_CallFunction(factory, "NdiOiOii",
+                                          PyOS_FSPath(database),
```
Comment (erlend-aasland) [inline Modules/_sqlite/module.c]: As you mention in https://github.com/python/cpython/pull/92926#issuecomment-1131475986, if `PyOS_FSPath` is not needed, we can revert this particular change.

File: Modules/_sqlite/connection.c [hunk 4]
```diff
@@ -188,17 +154,22 @@ _sqlite3.Connection.__init__ as pysqlite_connection_init
...
+/*[clinic end generated code: output=839eb2fee4293bda input=b8ce63dc6f70a383]*/
 {
     if (PySys_Audit("sqlite3.connect", "s", database) < 0) {
         return -1;
     }
 
+    PyObject *bytes;
+    if (!PyUnicode_FSConverter(database, &bytes)) {
+        return -1;
+    }
```
Comment (erlend-aasland) [inline Modules/_sqlite/connection.c]: Regarding https://github.com/python/cpython/pull/92926#issuecomment-1131475986, I think a short comment above line 168 on why this need to come after `PySys_Audit` would be nice.

**Review by erlend-aasland — APPROVED** [review]
Looks good; thanks a lot!

I left some minor comments.

File: Lib/test/test_sqlite3/test_dbapi.py (hunk 2 above)
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_dbapi.py]: Oops, my comment is outdated already.

File: Lib/test/test_sqlite3/test_dbapi.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/test/test_sqlite3/test_dbapi.py:664]: It unlinks TESTFN, not TESTFN_UNDECODABLE.

File: Modules/_sqlite/connection.c (hunk 4 above)
Comment (serhiy-storchaka) [reply Modules/_sqlite/connection.c]: I do not have a good comment which would survive the time. In any case, `test_open_with_undecodable_path` fails without these changes, so it is unlikely that somebody will incidentally reintroduce the same bug.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Good point.

File: Lib/test/test_sqlite3/test_dbapi.py (hunk 1 above)
Comment (erlend-aasland) [reply Lib/test/test_sqlite3/test_dbapi.py:664]: Darn, that is correct.

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
The tree buildbot failures so far all seem unrelated to this PR.

- AMD64 Windows10 PR
- s390x Fedora LTO + PGO PR
- s390x Fedora LTO PR

**Comment by serhiy-storchaka** [comment]
I ran tests on buildbots because the issue is locale-specific.

**Comment by erlend-aasland** [comment]
BTW, if you want me to land this, add me as assigned.

**Pull request merged**